File: М11_ClusterVMS/clustervms/cluster/apphost.py
"""The Node under a scheduler: М10's AppHost plus what М11 adds.

    prologue      identity → migrate → rehydrate → epoch by CAS → lease
    publish()     every second: publish on change, with a floor           (Lesson 3)
    lease()       renew by reading my epoch; fence myself if it moved      (Lesson 4)
    heartbeat()   a timestamp in my Variable, for node_failover_seconds    (Lesson 4)

Everything М10 does — reconcile, pump_buses, report, retention, console —
is inherited unchanged. Only the actuator's gate and the report grow:
a fenced instance starts nothing, and `replicated` joins the conditions.
"""
from __future__ import annotations

import asyncio
import dataclasses
import logging
import os
import signal
import time

from . import identity as identity_mod
from .directory import Directory
from .epoch import Lease, next_epoch
from .objectstore import ObjectStore
from .publish import Publisher
from .rehydrate import RestoreResult, rehydrate
from .reindex import RealFs as ReindexFs, sweep as reindex_sweep
from .variables import Variables

from apphost.apphost import MIGRATIONS, AppHost   # М10's nodevms, on sys.path via cluster/__init__.py

log = logging.getLogger("cluster.apphost")


class ClusterAppHost(AppHost):
    def __init__(self, settings, store, vars_: Variables, objects: ObjectStore,
                 identity: identity_mod.Identity, actuator=None, key=None,
                 lease_ttl: float = 30.0, lease_margin: float = 5.0,
                 publish_floor: float = 5.0, heartbeat_interval: float = 30.0,
                 clock=time.monotonic, wall=time.time):
        super().__init__(settings, store, actuator, key)
        self.vars, self.objects, self.identity = vars_, objects, identity
        self.lease_ttl, self.lease_margin = lease_ttl, lease_margin
        self.heartbeat_interval = heartbeat_interval
        self.clock, self.wall = clock, wall
        self.publisher = Publisher(identity.node, store, vars_, objects, publish_floor, clock)
        self.directory = Directory(vars_, clock=clock)
        self.lease: Lease | None = None
        self.failover: dict = {}
        self.replicated_now = False
        self.restore: RestoreResult | None = None
        self._resume_from_ts: float | None = None

    # -- prologue: the six steps ---------------------------------------------
    async def prologue(self) -> RestoreResult:
        node = self.identity.node
        self.restore = await rehydrate(self.identity, self.store, self.objects)     # steps 2–4
        self.publisher.published_rev = self.identity.config_revision
        hb, _ = self.vars.get(f"nodes/{node}/heartbeat")
        epoch, idx = next_epoch(self.vars, node)                                   # step 5
        self.settings = dataclasses.replace(self.settings, epoch=epoch)            # step 6: EPOCH in the path
        self.actuator.settings = self.settings           # the epoch in every new segment path
        self.lease = Lease(self.vars, node, epoch, self.lease_ttl, self.lease_margin, self.clock)
        fo, _ = self.vars.get(f"nodes/{node}/failover")
        self.failover = {k: float(v) for k, v in (fo or {}).items()}
        if self.restore.state == "restored" and hb and hb.get("ts"):
            self._resume_from_ts = float(hb["ts"])       # the old instance's last sign of life
        log.info("%s: %s (rev %d, %d cameras); epoch %d (ModifyIndex %d)", node, self.restore.state,
                 self.restore.revision, self.restore.cameras, epoch, idx)
        return self.restore

    # -- gates ----------------------------------------------------------------
    def _actuate(self, verb: str, cam: dict) -> bool:
        if verb in ("start", "restart") and self.lease is not None and not self.lease.may_write():
            return False                                 # fenced, or the lease ran out: start nothing
        return super()._actuate(verb, cam)

    async def reconcile_once(self):
        actions = await super().reconcile_once()
        if self._resume_from_ts is not None and any(v == "start" for v, _ in actions):
            secs = max(0.0, self.wall() - self._resume_from_ts)
            self.failover = {"last": secs, "worst": max(secs, self.failover.get("worst", 0.0))}
            self._resume_from_ts = None
            try:
                _, idx = self.vars.get(f"nodes/{self.identity.node}/failover")
                self.vars.put(f"nodes/{self.identity.node}/failover", self.failover, cas=idx)
            except Exception:                            # noqa: BLE001
                log.exception("could not record failover time")
            log.warning("%s: recording resumed %.1fs after the old instance's last heartbeat (worst %.1fs)",
                        self.identity.node, secs, self.failover["worst"])
        return actions

    async def report_once(self) -> None:
        await super().report_once()
        rev = await self.store.config_revision()
        ok, reason = self.publisher.replicated(rev)
        self.replicated_now = ok
        for cam in self.desired.rows:
            await self.store.set_condition(cam["id"], "replicated", ok, reason)

    # -- the added tasks ------------------------------------------------------
    async def publish(self) -> None:
        while not self.stopping.is_set():
            try:
                await self.publisher.publish_once()
            except Exception:                            # noqa: BLE001
                log.exception("publish task failed; retrying")
            await asyncio.sleep(1.0)

    def fence(self, why: str) -> None:
        if not self.recording_allowed:
            return
        log.error("%s: FENCED (%s). Stopping every pipeline; this instance's epoch %d is stale.",
                  self.identity.node, why, self.settings.epoch)
        self.recording_allowed = False
        self.actuator.stop_all()
        self.reconciler.actual.clear()
        self.wake.set()

    async def lease_task(self) -> None:
        interval = max(1.0, (self.lease_ttl - self.lease_margin) / 3)
        while not self.stopping.is_set():
            if self.lease is not None:
                if not self.lease.renew():
                    self.fence("a newer epoch was issued" if self.lease.fenced else "lease expired without renewal")
            await asyncio.sleep(interval)

    async def reindex(self, interval: float = 600.0) -> None:
        """After a restore and periodically: files back into rows — the
        fenced instance's footage, and a returned server's archive."""
        fs = ReindexFs()
        while not self.stopping.is_set():
            try:
                rep = await reindex_sweep(self.store, fs, self.settings.archive_dir, self.settings.epoch,
                                          self.settings.segment_seconds)
                if rep.reindexed:
                    log.info("reindex: %d segments, %d from fenced epochs", rep.reindexed, rep.fenced)
            except Exception:                            # noqa: BLE001
                log.exception("reindex sweep failed")
            await asyncio.sleep(interval)

    async def heartbeat(self) -> None:
        while not self.stopping.is_set():
            try:
                if self.lease is not None and self.lease.may_write():
                    _, idx = self.vars.get(f"nodes/{self.identity.node}/heartbeat")
                    self.vars.put(f"nodes/{self.identity.node}/heartbeat",
                                  {"ts": f"{self.wall():.3f}", "epoch": self.settings.epoch}, cas=idx)
            except Exception:                            # noqa: BLE001
                log.debug("heartbeat write failed (cluster unreachable?)")
            await asyncio.sleep(self.heartbeat_interval)

    # -- lifecycle --------------------------------------------------------------
    async def run(self, serve_console: bool = True) -> None:
        ok = await self.store.migrate(MIGRATIONS)                                   # step 1
        if not ok:
            log.error("running on the previous schema; the box keeps recording")
        await self.prologue()
        listener = None
        try:
            listener = await self.store.listen("cameras", lambda _payload: self.wake.set())
        except Exception:                                                          # noqa: BLE001
            log.exception("LISTEN failed; polling alone (slower, still correct)")
        loop = asyncio.get_running_loop()
        for sig in (signal.SIGTERM, signal.SIGINT):
            loop.add_signal_handler(sig, self.stopping.set)
        tasks = [asyncio.create_task(self.reconcile(), name="reconcile"),
                 asyncio.create_task(self.pump_buses(), name="pump_buses"),
                 asyncio.create_task(self.report(), name="report"),
                 asyncio.create_task(self.retention(), name="retention"),
                 asyncio.create_task(self.publish(), name="publish"),
                 asyncio.create_task(self.lease_task(), name="lease"),
                 asyncio.create_task(self.heartbeat(), name="heartbeat"),
                 asyncio.create_task(self.reindex(), name="reindex")]
        if serve_console:
            import uvicorn
            from console.app import create_app
            from .console import add_cluster_routes
            app = create_app(self)
            add_cluster_routes(app, self)
            server = uvicorn.Server(uvicorn.Config(app, host=self.settings.console_host,
                                                   port=self.settings.console_port, log_level="warning"))
            tasks.append(asyncio.create_task(server.serve(), name="console"))
        await self.stopping.wait()
        log.info("stopping: closing pipelines (each finalizes its open segment)")
        self.actuator.stop_all()
        for t in tasks:
            t.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        try:
            await self.publisher.publish_once()          # a last publish, best effort
            await self.report_once()
        finally:
            if listener is not None:
                await self.store.pool.release(listener)


async def main() -> None:
    from apphost.config import Settings
    from apphost.secrets import ColumnKey
    from apphost.store import PgStore
    from .configio import ClusterStoreMixin
    from .objectstore import open_store
    from .variables import NomadVariables

    logging.basicConfig(level=os.environ.get("LOG_LEVEL", "INFO"),
                        format="%(asctime)s %(name)s %(levelname)s %(message)s")

    class ClusterPgStore(ClusterStoreMixin, PgStore):
        pass

    vars_ = NomadVariables()
    ident = identity_mod.from_environment(vars_)
    settings = Settings()
    key = None
    if ident.column_key:
        os.makedirs(os.path.dirname(settings.column_key_file), exist_ok=True)
        with open(settings.column_key_file, "wb") as f:
            f.write(ident.column_key)               # М10's key, delivered by the cluster, not from a backup
        key = ColumnKey(ident.column_key)
    objects = open_store(os.environ.get("OBJECT_STORE_URL", "file:///data/restore"))
    store = await ClusterPgStore.connect(settings.database_url)
    try:
        host = ClusterAppHost(settings, store, vars_, objects, ident, key=key,
                              lease_ttl=float(os.environ.get("LEASE_TTL", "30")),
                              lease_margin=float(os.environ.get("LEASE_MARGIN", "5")),
                              publish_floor=float(os.environ.get("PUBLISH_FLOOR", "5")),
                              heartbeat_interval=float(os.environ.get("HEARTBEAT_INTERVAL", "30")))
        await host.run()
    finally:
        await store.close()

File: М11_ClusterVMS/clustervms-go/cluster/apphost.go
package cluster

// The Node under a scheduler: М10's AppHost plus what М11 adds.
//
//	Prologue      identity → migrate → rehydrate → epoch by CAS → lease
//	publish       every second: publish on change, with a floor           (Lesson 3)
//	lease         renew by reading my epoch; fence myself if it moved      (Lesson 4)
//	heartbeat     a timestamp in an OBJECT, for node_failover_seconds      (Lesson 4)
//	reindex       files back into rows: the fenced instance's footage      (Lesson 4)
//
// What М10 does — reconcile, report — is the same loop (nodevms/reconciler,
// imported) driven by goroutines and tickers instead of asyncio tasks.
// Only the actuator's gate and the report grow: a fenced instance starts
// nothing, and `replicated` joins the conditions.

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"sync"
	"time"

	"nodevms/reconciler"
)

type desired struct {
	mu   sync.Mutex
	rows []CameraRow
	byID map[int64]CameraRow
}

func (d *desired) set(rows []CameraRow) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.rows = rows
	d.byID = make(map[int64]CameraRow, len(rows))
	for _, r := range rows {
		d.byID[r.ID] = r
	}
}

func (d *desired) Desired() []reconciler.Camera {
	d.mu.Lock()
	defer d.mu.Unlock()
	out := make([]reconciler.Camera, 0, len(d.rows))
	for _, r := range d.rows {
		out = append(out, reconciler.Camera{ID: r.ID, Enabled: r.Enabled, Revision: r.Revision})
	}
	return out
}

func (d *desired) Rows() []CameraRow {
	d.mu.Lock()
	defer d.mu.Unlock()
	return append([]CameraRow(nil), d.rows...)
}

type ClusterAppHost struct {
	Settings Settings
	Store    Store
	Vars     Variables
	Objects  ObjectStore
	Identity Identity
	Actuator Actuator
	Clock    Clock          // monotonic
	Wall     func() float64 // unix seconds

	Desired          *desired
	Reconciler       *reconciler.Reconciler
	Publisher        *Publisher
	Directory        *Directory
	Lease            *Lease
	Restore          *RestoreResult
	Failover         map[string]float64 // last, worst — node_failover_seconds
	RecordingAllowed bool
	ReplicatedNow    bool
	Passes           int

	mu           sync.Mutex // guards the loop-shared fields above during Run
	startedAt    float64
	resumeFromTS float64 // > 0 while a restore waits for its first start
	wake         chan struct{}
}

func NewClusterAppHost(settings Settings, store Store, v Variables, objects ObjectStore, id Identity, act Actuator,
	clock Clock, wall func() float64) *ClusterAppHost {
	if clock == nil {
		clock = Monotonic()
	}
	if wall == nil {
		wall = func() float64 { return float64(time.Now().UnixNano()) / 1e9 }
	}
	h := &ClusterAppHost{Settings: settings, Store: store, Vars: v, Objects: objects, Identity: id, Actuator: act,
		Clock: clock, Wall: wall, Desired: &desired{}, Failover: map[string]float64{}, RecordingAllowed: true,
		wake: make(chan struct{}, 1)}
	h.Reconciler = reconciler.New(h.Desired, h.actuate)
	h.Reconciler.MaxBackoff, h.Reconciler.StallFailures = settings.MaxBackoff, settings.StallFailures
	h.Publisher = NewPublisher(id.Node, store, v, objects, settings.PublishFloor, clock)
	h.Directory = NewDirectory(v, 5, clock)
	h.startedAt = clock()
	act.SetSettings(settings)
	return h
}

// -- prologue: the six steps ------------------------------------------------

func (h *ClusterAppHost) Prologue() (RestoreResult, error) {
	node := h.Identity.Node
	r, err := Rehydrate(h.Identity, h.Store, h.Objects) // steps 2–4
	if err != nil {
		return r, err
	}
	h.Restore = &r
	h.Publisher.PublishedRev = h.Identity.ConfigRevision
	hb := h.readHeartbeat()
	epoch, idx, err := NextEpoch(h.Vars, node) // step 5
	if err != nil {
		return r, err
	}
	h.Settings.Epoch = epoch           // step 6: the epoch in the path
	h.Actuator.SetSettings(h.Settings) // in every new segment path
	h.Lease = NewLease(h.Vars, node, epoch, h.Settings.LeaseTTL, h.Settings.LeaseMargin, h.Clock)
	if fo, _, err := h.Vars.Get("nodes/" + node + "/failover"); err == nil {
		for k, v := range fo {
			h.Failover[k], _ = strconv.ParseFloat(v, 64)
		}
	}
	if r.State == "restored" && hb != nil && hb.TS > 0 {
		h.resumeFromTS = hb.TS // the old instance's last sign of life
	}
	log.Printf("%s: %s (rev %d, %d cameras); epoch %d (ModifyIndex %d)", node, r.State, r.Revision, r.Cameras, epoch, idx)
	return r, nil
}

// -- gates ----------------------------------------------------------------------

func (h *ClusterAppHost) actuate(verb string, cam reconciler.Camera) bool {
	if verb == "start" || verb == "restart" {
		if !h.RecordingAllowed { // storage unavailable is a REASON, not a phase
			return false
		}
		if h.Lease != nil && !h.Lease.MayWrite() { // fenced, or the lease ran out: start nothing
			return false
		}
	}
	h.Desired.mu.Lock()
	row := h.Desired.byID[cam.ID]
	h.Desired.mu.Unlock()
	row.ID = cam.ID
	return h.Actuator.Actuate(verb, row)
}

func (h *ClusterAppHost) now() float64 { return h.Clock() - h.startedAt }

func (h *ClusterAppHost) ReconcileOnce() ([]reconciler.Action, error) {
	rows, err := h.Store.FetchDesired()
	if err != nil {
		return nil, err
	}
	h.Desired.set(rows)
	h.mu.Lock()
	defer h.mu.Unlock()
	actions := h.Reconciler.Reconcile(h.now())
	h.Passes++
	started := false
	for _, a := range actions {
		log.Printf("reconcile: %s camera %d", a.Verb, a.ID)
		started = started || a.Verb == "start"
	}
	if h.resumeFromTS > 0 && started {
		secs := h.Wall() - h.resumeFromTS
		if secs < 0 {
			secs = 0
		}
		worst := h.Failover["worst"]
		if secs > worst {
			worst = secs
		}
		h.Failover = map[string]float64{"last": secs, "worst": worst}
		h.resumeFromTS = 0
		path := "nodes/" + h.Identity.Node + "/failover"
		if _, idx, err := h.Vars.Get(path); err == nil {
			_, err = h.Vars.Put(path, Items{"last": fmtF(secs), "worst": fmtF(worst)}, idx)
			if err != nil {
				log.Printf("could not record failover time: %v", err)
			}
		}
		log.Printf("%s: recording resumed %.1fs after the old instance's last heartbeat (worst %.1fs)", h.Identity.Node, secs, worst)
	}
	return actions, nil
}

func fmtF(f float64) string { return strconv.FormatFloat(f, 'f', 1, 64) }

// phases: camera id -> (observed_revision, phase). Phase is a POSITION.
func (h *ClusterAppHost) phases() []StatusRow {
	actual, failures := h.Reconciler.Actual(), h.Reconciler.Failures()
	var out []StatusRow
	for _, cam := range h.Desired.Rows() {
		if !cam.Enabled {
			out = append(out, StatusRow{cam.ID, cam.Revision, "pending"})
			continue
		}
		have, running := actual[cam.ID]
		phase := "pending"
		switch {
		case running:
			phase = "running"
		case failures[cam.ID].N > 0:
			phase = "failed"
		}
		out = append(out, StatusRow{cam.ID, have, phase})
	}
	return out
}

func (h *ClusterAppHost) ReportOnce() error {
	h.mu.Lock()
	defer h.mu.Unlock()
	if err := h.Store.Report(h.phases()); err != nil {
		return err
	}
	rev, err := h.Store.ConfigRevision()
	if err != nil {
		return err
	}
	ok, reason := h.Publisher.Replicated(rev)
	h.ReplicatedNow = ok
	failures := h.Reconciler.Failures()
	for _, cam := range h.Desired.Rows() {
		f, failing := failures[cam.ID]
		fr := ""
		if failing {
			fr = fmt.Sprintf("failing, retry in %.0fs", f.Delay)
		}
		sr := ""
		if !h.RecordingAllowed {
			sr = "disk full; policy=stop_recording"
		}
		for _, c := range []struct {
			name   string
			ok     bool
			reason string
		}{{"camera_reachable", !failing, fr}, {"storage_available", h.RecordingAllowed, sr},
			{"licensed", true, ""}, {"replicated", ok, reason}} {
			if err := h.Store.SetCondition(cam.ID, c.name, c.ok, c.reason); err != nil {
				return err
			}
		}
	}
	return nil
}

// Fence stops everything: this instance's epoch is stale.
func (h *ClusterAppHost) Fence(why string) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if !h.RecordingAllowed {
		return
	}
	log.Printf("%s: FENCED (%s). Stopping every pipeline; this instance's epoch %d is stale.", h.Identity.Node, why, h.Settings.Epoch)
	h.RecordingAllowed = false
	h.Actuator.StopAll()
	h.Reconciler.Clear()
	h.Wake()
}

func (h *ClusterAppHost) Wake() {
	select {
	case h.wake <- struct{}{}:
	default:
	}
}

// -- the heartbeat is an object ------------------------------------------------
//
// Small and FREQUENT and never queried — the three-stores rule (Lesson 2)
// says that is not raft. A thousand Nodes writing a raft entry every ten
// seconds is a hundred commits a second replicated to every server; a
// thousand tiny objects a second is nothing to an object store.

type heartbeat struct {
	TS       float64        `json:"ts"`
	Epoch    int64          `json:"epoch"`
	Revision int64          `json:"revision,omitempty"`
	Server   string         `json:"server,omitempty"`
	Cameras  []cameraStatus `json:"cameras,omitempty"`
}

// cameraStatus is the Node's own /status row, carried in the heartbeat so
// the cluster console builds the camera list without calling any Node
// (М12, "The camera list"). Server lets a dead server read as one cause.
type cameraStatus struct {
	ID               int64  `json:"id"`
	Name             string `json:"name"`
	Site             string `json:"site"`
	Enabled          bool   `json:"enabled"`
	Phase            string `json:"phase"`
	Revision         int64  `json:"revision"`
	ObservedRevision int64  `json:"observed_revision"`
}

// HeartbeatPayload is the object the heartbeat task writes.
func (h *ClusterAppHost) HeartbeatPayload() heartbeat {
	rows := h.Desired.Rows()
	phases := map[int64]StatusRow{}
	for _, p := range h.phases() {
		phases[p.Camera] = p
	}
	cams := make([]cameraStatus, 0, len(rows))
	for _, r := range rows {
		p := phases[r.ID]
		if p.Phase == "" {
			p.Phase = "pending"
		}
		cams = append(cams, cameraStatus{r.ID, r.Name, r.SiteID, r.Enabled, p.Phase, r.Revision, p.ObservedRevision})
	}
	server := os.Getenv("NOMAD_NODE_ID")
	if server == "" {
		server, _ = os.Hostname()
	}
	return heartbeat{h.Wall(), h.Settings.Epoch, h.Identity.ConfigRevision, server, cams}
}

func (h *ClusterAppHost) readHeartbeat() *heartbeat {
	raw, err := h.Objects.Get(h.Identity.Node + "/heartbeat")
	if err != nil || len(raw) == 0 {
		return nil
	}
	var hb heartbeat
	if json.Unmarshal(raw, &hb) != nil {
		return nil
	}
	return &hb
}

func (h *ClusterAppHost) heartbeatOnce() {
	if h.Lease == nil || !h.Lease.MayWrite() {
		return
	}
	b, _ := json.Marshal(h.HeartbeatPayload())
	if err := h.Objects.Put(h.Identity.Node+"/heartbeat", b); err != nil {
		log.Printf("heartbeat write failed (object store unreachable?): %v", err)
	}
}

func (h *ClusterAppHost) leaseOnce() {
	if h.Lease == nil {
		return
	}
	if !h.Lease.Renew() {
		why := "lease expired without renewal"
		if h.Lease.Fenced {
			why = "a newer epoch was issued"
		}
		h.Fence(why)
	}
}

func (h *ClusterAppHost) reindexOnce() {
	rep, err := Sweep(h.Store, RealFs{}, h.Settings.ArchiveDir, h.Settings.Epoch, h.Settings.SegmentSeconds, h.Wall())
	if err != nil {
		log.Printf("reindex sweep failed: %v", err)
	} else if rep.Reindexed > 0 {
		log.Printf("reindex: %d segments, %d from fenced epochs", rep.Reindexed, rep.Fenced)
	}
}

// -- lifecycle ---------------------------------------------------------------

func every(ctx context.Context, wg *sync.WaitGroup, interval time.Duration, wake <-chan struct{}, f func()) {
	wg.Add(1)
	go func() {
		defer wg.Done()
		t := time.NewTicker(interval)
		defer t.Stop()
		for {
			f()
			select {
			case <-ctx.Done():
				return
			case <-t.C:
			case <-wake:
			}
		}
	}()
}

// Run is the process: migrate, the prologue, then one goroutine per
// concern until ctx is cancelled. The timer is correctness; the wake is
// latency.
func (h *ClusterAppHost) Run(ctx context.Context, serveConsole bool) error {
	if ok, err := h.Store.Migrate(""); err != nil {
		return err
	} else if !ok {
		log.Printf("running on the previous schema; the box keeps recording")
	}
	if _, err := h.Prologue(); err != nil {
		return err
	}
	sec := func(f float64) time.Duration { return time.Duration(f * float64(time.Second)) }
	var wg sync.WaitGroup
	never := make(chan struct{})
	every(ctx, &wg, sec(h.Settings.PollInterval), h.wake, func() {
		if _, err := h.ReconcileOnce(); err != nil {
			log.Printf("reconcile pass failed; will retry: %v", err)
		}
	})
	every(ctx, &wg, sec(h.Settings.ReportInterval), never, func() {
		if err := h.ReportOnce(); err != nil {
			log.Printf("report failed; convergence continues regardless: %v", err)
		}
	})
	every(ctx, &wg, time.Second, never, func() { h.Publisher.PublishOnce() })
	interval := (h.Settings.LeaseTTL - h.Settings.LeaseMargin) / 3
	if interval < 1 {
		interval = 1
	}
	every(ctx, &wg, sec(interval), never, h.leaseOnce)
	every(ctx, &wg, sec(h.Settings.HeartbeatInterval), never, h.heartbeatOnce)
	every(ctx, &wg, sec(h.Settings.ReindexInterval), never, h.reindexOnce)
	if serveConsole {
		srv := ServeConsole(h, h.Settings.ConsoleAddr)
		go func() { <-ctx.Done(); srv.Close() }()
	}
	<-ctx.Done()
	log.Printf("stopping: closing pipelines (each finalizes its open segment)")
	h.Actuator.StopAll()
	wg.Wait()
	h.Publisher.PublishOnce() // a last publish, best effort
	return h.ReportOnce()
}

func sortInt64s(a []int64) { sort.Slice(a, func(i, j int) bool { return a[i] < a[j] }) }
